### PR TITLE
[javanaut][operationExpr][2/2]feat: Add Assignment operation expr with xor assignemnt

### PR DESCRIPTION
**Pull request by summer-ji-eng**
Enables ^= operator and add unit test.

### CODE REVIEW

File: src/main/java/com/google/api/generator/engine/ast/AssignmentOperationExpr.java [hunk 1]
```diff
@@ -130,9 +129,26 @@ private boolean isValidMultiplyAssignmentType(TypeNode variableType, TypeNode va
...
     private boolean isValidXORAssignmentType(TypeNode variableType, TypeNode valueType) {
-      return true;
+      // LHS is boolean or its boxed type, RHS should be boolean or its boxed type.
+      if (variableType.equals(TypeNode.BOOLEAN)) {
+        return valueType.equals(variableType);
+      }
+      // LHS is non-floating-point numeric types, RHS should be non-float-point numeric types or
+      // their boxed types.
+      if (TypeNode.isNumericType(variableType)
+          && !TypeNode.isBoxedType(variableType)
```
Comment (miraleung) [inline src/main/java/com/google/api/generator/engine/ast/AssignmentOperationExpr.java:142]: Consider checking up-front and returning false first if:
- One of `variableType` or `valueType`  is not a numeric type.
- `valueType` is boxed or is a floating-point type.
- `valueType` is a floating-point type.

**Review by miraleung — CHANGES_REQUESTED** [review]
LGTM with nits addressed.

File: src/main/java/com/google/api/generator/engine/ast/AssignmentOperationExpr.java [hunk 2]
```diff
@@ -130,9 +129,26 @@ private boolean isValidMultiplyAssignmentType(TypeNode variableType, TypeNode va
       return false;
     }
 
-    // TODO(summerji): Complete the type-checking for ^= and unit test.
+    // isValidXORAssignmentType validates the types for LHS variable expr and RHS expr.
+    // ^= can be applied on boolean and non-floating-point numeric type.
     private boolean isValidXORAssignmentType(TypeNode variableType, TypeNode valueType) {
```
[the remark this answers is not in the document]
Comment (summer-ji-eng) [reply src/main/java/com/google/api/generator/engine/ast/AssignmentOperationExpr.java:134]: sg.

File: src/main/java/com/google/api/generator/engine/ast/AssignmentOperationExpr.java (hunk 1 above)
Comment (summer-ji-eng) [reply src/main/java/com/google/api/generator/engine/ast/AssignmentOperationExpr.java:142]: I could do like 
```
    private boolean isValidXorAssignmentType(TypeNode variableType, TypeNode valueType) {

      // LHS is boolean or its boxed type, RHS should be boolean or its boxed type.
      if (variableType.equals(TypeNode.BOOLEAN)) {
        return valueType.equals(variableType);
      }
      // LHS is integer boxed type, RHS should be non-floating-point numeric types or their boxed
      // types.
      if (variableType.equals(TypeNode.INT)) {
        return TypeNode.isNumericType(valueType) && !TypeNode.isFloatingPointType(valueType);
      }

      if (!TypeNode.isNumericType(variableType) || !TypeNode.isNumericType(valueType)) {
        return false;
      }
      if (TypeNode.isFloatingPointType(variableType) || TypeNode.isFloatingPointType(valueType)) {
        return false;
      }
      if (TypeNode.isBoxedType(variableType)) {
        return false;
      }
      return true;
    }
```
But I don't see the point of checking up-front. Could you explain more the benefit to do so?

Comment (miraleung) [reply src/main/java/com/google/api/generator/engine/ast/AssignmentOperationExpr.java:142]: When rearranged this way, we can now simplify it as follows:
```
return TypeNode.isNumericType(variableType) && TypeNode.isNumericType(valueType)
    && !TypeNode.isFloatingPointType(variableType) && !TypeNode.isFloatingPointType(valueType)
    && !TypeNode.isBoxedType(variableType);
```

The original benefit was to make the true-checks easier to read, but this rearrangement works too.

File: src/test/java/com/google/api/generator/engine/ast/AssignmentOperationExprTest.java [hunk 3]
```diff
@@ -644,8 +644,594 @@ public void invalidMultiplyAndAssignmentOperationExpr_voidType() {
...
+    // No valid swap case.
+    VariableExpr lhsExpr = createVariableExpr(TypeNode.BOOLEAN, "x");
+    ValueExpr rhsExpr = ValueExpr.withValue(NullObjectValue.create());
+    assertThrows(
+        IllegalStateException.class,
+        () -> AssignmentOperationExpr.xorAssignmentWithExprs(lhsExpr, rhsExpr));
+  }
+
+  @Test
+  public void invalidORAssignmentOperationExpr_booleanWithNewObjectTypes() {
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/AssignmentOperationExprTest.java:725]: Nit: `invalidXor`

File: src/test/java/com/google/api/generator/engine/ast/AssignmentOperationExprTest.java [hunk 4]
```diff
@@ -644,8 +644,594 @@ public void invalidMultiplyAndAssignmentOperationExpr_voidType() {
...
+    // Swap case in "invalidORAssignmentOperationExpr_floatWithCharTypes".
+    VariableExpr lhsExpr = createVariableExpr(TypeNode.CHAR, "x");
+    VariableExpr rhsExpr = createVariableExpr(TypeNode.FLOAT, "y");
+    assertThrows(
+        IllegalStateException.class,
+        () -> AssignmentOperationExpr.xorAssignmentWithExprs(lhsExpr, rhsExpr));
+  }
+
+  @Test
+  public void invalidORAssignmentOperationExpr_charWithFloatBoxedTypes() {
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/AssignmentOperationExprTest.java:874]: `invalidXor`?

Comment (miraleung) [reply src/test/java/com/google/api/generator/engine/ast/AssignmentOperationExprTest.java:874]: Same below.

File: src/test/java/com/google/api/generator/engine/ast/AssignmentOperationExprTest.java [hunk 5]
```diff
@@ -644,8 +644,594 @@ public void invalidMultiplyAndAssignmentOperationExpr_voidType() {
...
+    // Swap case in invalidORAssignmentOperationExpr_referenceWithIntegerBoxedTypes.
+    VariableExpr lhsExpr = createVariableExpr(TypeNode.INT_OBJECT, "x");
+    VariableExpr rhsExpr = createVariableExpr(TypeNode.STRING, "y");
+    assertThrows(
+        IllegalStateException.class,
+        () -> AssignmentOperationExpr.xorAssignmentWithExprs(lhsExpr, rhsExpr));
+  }
+
+  @Test
+  public void invalidORAssignmentOperationExpr_integerBoxedWithNewObjectTypes() {
```
Comment (miraleung) [inline src/test/java/com/google/api/generator/engine/ast/AssignmentOperationExprTest.java:1111]: Ditto for the others in this file.

**Review by miraleung — CHANGES_REQUESTED** [review]

File: src/test/java/com/google/api/generator/engine/ast/AssignmentOperationExprTest.java (hunk 3 above)
Comment (summer-ji-eng) [reply src/test/java/com/google/api/generator/engine/ast/AssignmentOperationExprTest.java:725]: oops. Copy and paste a lot. Fixed right now. Should be good.

File: src/main/java/com/google/api/generator/engine/ast/AssignmentOperationExpr.java (hunk 1 above)
Comment (summer-ji-eng) [reply src/main/java/com/google/api/generator/engine/ast/AssignmentOperationExpr.java]: Updated, LMK what you think on new version.

**Review by miraleung — APPROVED** [review]

**Pull request merged**